### PR TITLE
`azurerm_frontdoor`: resources sorting part 2

### PR DESCRIPTION
**Pull request by favoretti**
Switch `TypeList` to `TypeSet`

### CODE REVIEW

**Comment by favoretti** [comment]
```
$ TF_ACC=1 go test -v ./azurerm/internal/services/frontdoor -timeout=1000m
=== RUN   TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
=== PAUSE TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
=== RUN   TestAccFrontDoorFirewallPolicy_basic
=== PAUSE TestAccFrontDoorFirewallPolicy_basic
=== RUN   TestAccFrontDoorFirewallPolicy_requiresImport
=== PAUSE TestAccFrontDoorFirewallPolicy_requiresImport
=== RUN   TestAccFrontDoorFirewallPolicy_update
=== PAUSE TestAccFrontDoorFirewallPolicy_update
=== RUN   TestAccFrontDoorFirewallPolicy_complete
=== PAUSE TestAccFrontDoorFirewallPolicy_complete
=== RUN   TestAccFrontDoor_basic
=== PAUSE TestAccFrontDoor_basic
=== RUN   TestAccFrontDoor_global
=== PAUSE TestAccFrontDoor_global
=== RUN   TestAccFrontDoor_requiresImport
=== PAUSE TestAccFrontDoor_requiresImport
=== RUN   TestAccFrontDoor_update
=== PAUSE TestAccFrontDoor_update
=== RUN   TestAccFrontDoor_multiplePools
=== PAUSE TestAccFrontDoor_multiplePools
=== RUN   TestAccFrontDoor_complete
=== PAUSE TestAccFrontDoor_complete
=== RUN   TestAccFrontDoor_waf
=== PAUSE TestAccFrontDoor_waf
=== RUN   TestAccFrontDoor_EnableDisableCache
=== PAUSE TestAccFrontDoor_EnableDisableCache
=== CONT  TestAccFrontDoorCustomHttpsConfiguration_CustomHttps
=== CONT  TestAccFrontDoor_requiresImport
=== CONT  TestAccFrontDoorFirewallPolicy_requiresImport
=== CONT  TestAccFrontDoorFirewallPolicy_complete
=== CONT  TestAccFrontDoor_global
=== CONT  TestAccFrontDoor_basic
=== CONT  TestAccFrontDoorFirewallPolicy_update
=== CONT  TestAccFrontDoor_waf
=== CONT  TestAccFrontDoor_EnableDisableCache
=== CONT  TestAccFrontDoor_multiplePools
=== CONT  TestAccFrontDoor_complete
=== CONT  TestAccFrontDoor_update
=== CONT  TestAccFrontDoorFirewallPolicy_basic
--- PASS: TestAccFrontDoorFirewallPolicy_complete (104.38s)
--- PASS: TestAccFrontDoorFirewallPolicy_basic (104.84s)
--- PASS: TestAccFrontDoorFirewallPolicy_requiresImport (116.86s)
--- PASS: TestAccFrontDoorFirewallPolicy_update (192.52s)
--- PASS: TestAccFrontDoor_global (263.40s)
--- PASS: TestAccFrontDoor_complete (263.44s)
--- PASS: TestAccFrontDoor_multiplePools (265.09s)
--- PASS: TestAccFrontDoor_waf (271.57s)
--- PASS: TestAccFrontDoor_requiresImport (273.58s)
--- PASS: TestAccFrontDoor_EnableDisableCache (520.26s)
--- PASS: TestAccFrontDoor_update (521.91s)
--- PASS: TestAccFrontDoor_basic (1367.15s)
--- PASS: TestAccFrontDoorCustomHttpsConfiguration_CustomHttps (2951.84s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/frontdoor	2953.798s
```

**Comment by favoretti** [comment]
@tombuildsstuff if you're ok with removing inline option completely, I'll clean up the commented out parts, otherwise I'll keep digging into fixing `TestAccFrontDoorCustomHttpsConfiguration_CustomHttps` with inline option still present.

Reasoning for removing it: it's marked as deprecated and we're breaking the current resource by using Sets anyway, so might as well bite the bullet on this one too.

File: azurerm/internal/services/frontdoor/frontdoor_resource.go [hunk 1]
```diff
@@ -443,27 +443,27 @@ func resourceFrontDoor() *schema.Resource {
...
-							Type:       schema.TypeList,
-							Optional:   true,
-							Computed:   true,
-							MaxItems:   1,
-							Deprecated: "Deprecated in favour of `azurerm_frontdoor_custom_https_configuration` resource",
-							Elem: &schema.Resource{
-								Schema: schemaCustomHttpsConfiguration(),
-							},
-						},
+						// "custom_https_configuration": {
```
Comment (timja) [inline azurerm/internal/services/frontdoor/frontdoor_resource.go:457]: we never fully dug into this but when we tried to use this resource we had an issue with the wrong certificate being picked, I suspect something to do with the ordering of the endpoints, so your PR(s) may fix this issue and allow us to upgrade to it.

If possible I would suggest maintaining it to allow people to upgrade in stages

Comment (favoretti) [reply azurerm/internal/services/frontdoor/frontdoor_resource.go:457]: I'll need to dive into why separate resource breaks then, something is fishy there.

**Comment by WodansSon** [comment]
@favoretti, let's follow up in teams as I have some concerns with this. I'll ping you later tonight US time/tomorrow morning UK time. 🙂

**Comment by timja** [comment]
> @favoretti, let's follow up in teams as I have some concerns with this. I'll ping you later tonight US time/tomorrow morning UK time. 🙂

Any progress here? I'm happy to help out if you need

File: azurerm/internal/services/frontdoor/frontdoor_resource.go [hunk 2]
```diff
@@ -344,7 +344,7 @@ func resourceFrontDoor() *schema.Resource {
 			},
 
 			"backend_pool": {
-				Type:     schema.TypeList,
+				Type:     schema.TypeSet,
 				MaxItems: 50,
```
Comment (timja) [inline azurerm/internal/services/frontdoor/frontdoor_resource.go:348]: removing this line will resolve the conflict with master
```suggestion
```

**Comment by favoretti** [comment]
@timja Unfortunately this PR and #10921 changed their relationship status to `It's complicated`. I can't make a decision on which one is better and someone from HC needs to make a call on which one we finalize and merge.

**Comment by ryanbowden** [comment]
@favoretti is there anyone way we can help ping someone at HC to get this looked at?

**Comment by favoretti** [comment]
@ryanbowden all interested parties are already present here :)

**Comment by timja** [comment]
Is it expected that I still get `diffs` when I haven't changed anything when testing out this branch?

I assume it's to do with 'dynamic' `frontend_endpoint`'s.

Interestingly enough it seems to be only the `frontend_endpoint` and no backend_pools etc that are trying to change.


<details>
<summary>Plan output</summary>

```hcldiff
Terraform will perform the following actions:

  # module.landing_zone.azurerm_frontdoor.main will be updated in-place
  ~ resource "azurerm_frontdoor" "main" {
        id                                           = "/subscriptions/b72ab7b7-723f-4b18-b6f6-03b0f2c6a1bb/resourceGroups/lz-sbox-rg/providers/Microsoft.Network/frontDoors/timj-sbox"
        name                                         = "timj-sbox"
        tags                                         = {}
        # (13 unchanged attributes hidden)




      - frontend_endpoint {
          - host_name                    = "timj-sbox.azurefd.net" -> null
          - id                           = "/subscriptions/b72ab7b7-723f-4b18-b6f6-03b0f2c6a1bb/resourceGroups/lz-sbox-rg/providers/Microsoft.Network/frontDoors/timj-sbox/frontendEndpoints/timj-sbox-azurefd-net" -> null
          - name                         = "timj-sbox-azurefd-net" -> null
          - session_affinity_enabled     = false -> null
          - session_affinity_ttl_seconds = 0 -> null
        }
      + frontend_endpoint {
          + host_name                    = "timj-sbox.azurefd.net"
          + id                           = "/subscriptions/b72ab7b7-723f-4b18-b6f6-03b0f2c6a1bb/resourceGroups/lz-sbox-rg/providers/Microsoft.Network/frontDoors/timj-sbox/frontendEndpoints/timj-sbox-azurefd-net"
          + name                         = "timj-sbox-azurefd-net"
          + session_affinity_enabled     = false
          + session_affinity_ttl_seconds = 0
        }
      - frontend_endpoint {
          - host_name                               = "timj.sandbox.platform.hmcts.net" -> null
          - id                                      = "/subscriptions/b72ab7b7-723f-4b18-b6f6-03b0f2c6a1bb/resourceGroups/lz-sbox-rg/providers/Microsoft.Network/frontDoors/timj-sbox/frontendEndpoints/timj" -> null
          - name                                    = "timj" -> null
          - session_affinity_enabled                = false -> null
          - session_affinity_ttl_seconds            = 0 -> null
          - web_application_firewall_policy_link_id = "/subscriptions/b72ab7b7-723f-4b18-b6f6-03b0f2c6a1bb/resourceGroups/lz-sbox-rg/providers/Microsoft.Network/frontDoorWebApplicationFirewallPolicies/timjsbox" -> null
        }
      + frontend_endpoint {
          + host_name                               = "timj.sandbox.platform.hmcts.net"
          + id                                      = (known after apply)
          + name                                    = "timj"
          + session_affinity_enabled                = false
          + session_affinity_ttl_seconds            = 0
          + web_application_firewall_policy_link_id = "/subscriptions/b72ab7b7-723f-4b18-b6f6-03b0f2c6a1bb/resourcegroups/lz-sbox-rg/providers/Microsoft.Network/frontdoorWebApplicationFirewallPolicies/timjsbox"
        }

        # (8 unchanged blocks hidden)
    }
```

</details>


I'm using this branch of our module to test this:

https://github.com/hmcts/terraform-module-frontdoor/tree/hack

and

<details>
<summary>main.tf</summary>

```hcl

variable "env" {
  default = "sbox"
}
data "azurerm_resource_group" "main" {
  name = "lz-${var.env}-rg"
}

data "azurerm_resource_group" "key_vault" {
  name = "core-infra-${var.env}-rg"
}

variable "frontends" {
  default = {}
}

data "azurerm_subscription" "current" {}

module "log_analytics_workspace" {
  source      = "git::https://github.com/hmcts/terraform-module-log-analytics-workspace-id.git?ref=master"
  environment = var.env
}


module "landing_zone" {
  source = "git::https://github.com/hmcts/terraform-module-frontdoor.git?ref=hack"

  providers = {
    azurerm      = azurerm
    azurerm.data = azurerm.data
  }

  common_tags                = {}
  env                        = var.env
  subscription               = "sandbox"
  project                    = "timj"
  location                   = "uksouth"
  frontends                  = var.frontends
  ssl_mode                   = "AzureKeyVault"
  resource_group             = data.azurerm_resource_group.main.name
  subscription_id            = data.azurerm_subscription.current.subscription_id
  certificate_key_vault_name = "cftapps-sbox"
  oms_env                    = "sandbox"
  certificate_name_check     = false
  key_vault_resource_group   = data.azurerm_resource_group.key_vault.name
  log_analytics_workspace_id = module.log_analytics_workspace.workspace_id

  add_access_policy = false
}

```

</details>

<details>
<summary>providers.tf</summary>

```hcl

provider "azurerm" {
  features {}
}

provider "azurerm" {
  alias = "data"
  subscription_id = "bf308a5c-0624-4334-8ff8-8dca9fd43783"
}

```

</details>

<details>
<summary>terraform.tfvars</summary>

frontends = [
  {
    enable_ssl = false
    product = "timj"
    name = "timj"
    custom_domain = "timj.sandbox.platform.hmcts.net"
    backend_domain = [
      "some-domain.com"]
    certificate_name = "STAR-sandbox-platform-hmcts-net"
    global_exclusions = []
  }
]

</details>

**Comment by timja** [comment]
I did some more experimenting, and found that my `web_application_firewall_policy_link_id` was changing in case which was causing the diff

**Comment by katbyte** [comment]
looks like @WodansSon as repo'd this and is working on a fix, so going to close this in favour of #11456

**Pull request closed without merging**